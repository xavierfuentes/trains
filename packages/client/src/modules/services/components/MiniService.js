import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styled from 'styled-components';
import moment from 'moment';

import { TIME_FORMAT } from '../constants/visualization';

const StyledMiniService = styled.article`
  display: flex;
  flex-direction: row;
  margin-bottom: 0.071em;
  background: #ffffff;
  padding: 0.714em;
  border-left: 0.357em solid transparent;

  .scheduled-time,
  .service-info,
  .service-realtime {
    padding: 0.357em;
  }

  .scheduled-time {
    flex: 1;
    font-weight: 700;
    font-size: 1.143em
  }

  .service-info {
    flex: 3;
  }

  .service-realtime {
    flex: 2;
    text-align: right
  }

  .service-identifier {

  }

  .service-operator {
    color: #9b9f9e
    font-size: 0.857em
  }

  &.active {
    background: #d8f3ec;
    border-left-color: #244774;
  }
`;

class MiniService extends React.PureComponent {
  static propTypes = {
    service: PropTypes.shape({
      realTimeUpdatesInfo: PropTypes.shape({
        realTimeServiceInfo: PropTypes.shape({
          realTime: PropTypes.string
        })
      }),
      serviceIdentifier: PropTypes.string.isRequired,
      scheduledInfo: PropTypes.shape({
        scheduledTime: PropTypes.string.isRequired,
        scheduledPlatform: PropTypes.string
      }).isRequired,
      serviceOperator: PropTypes.string.isRequired
    }),
    activateService: PropTypes.func
  };

  handleClick = () => {
    const { service, activateService } = this.props;
    activateService && activateService(service);
  };

  render() {
    const { active, service } = this.props;
    const isActive = active && service.serviceIdentifier === active.serviceIdentifier;
    const scheduledTime = moment(service.scheduledInfo.scheduledTime);
    const realTime =
      service.realTimeUpdatesInfo &&
      service.realTimeUpdatesInfo.realTimeServiceInfo &&
      moment(service.realTimeUpdatesInfo.realTimeServiceInfo.realTime);
    const formattedScheduledTime = scheduledTime.format(TIME_FORMAT);
    const formattedRealTime = realTime && realTime.format(TIME_FORMAT);
    const isOnTime = realTime && (realTime.isAfter(scheduledTime) || realTime.isSame(scheduledTime));
    const isDelayed =
      service.realTimeUpdatesInfo &&
      service.realTimeUpdatesInfo.realTimeServiceInfo &&
      service.realTimeUpdatesInfo.realTimeServiceInfo.realTime;
    const isCancelled =
      service.realTimeUpdatesInfo &&
      service.realTimeUpdatesInfo.cancelled &&
      service.realTimeUpdatesInfo.cancelled.isCancelled;
    return (
      <StyledMiniService onClick={this.handleClick} className={classNames('service', { active: isActive })}>
        <div className="scheduled-time">{formattedScheduledTime}</div>

        <div className="service-info">
          <div className="service-identifier">{service.serviceIdentifier}</div>
          <div className="service-operator">{service.serviceOperator}</div>
        </div>

        <div className="service-realtime">
          {service.realTimeUpdatesInfo &&
            service.realTimeUpdatesInfo.realTimeServiceInfo && (
              <div className="service-platform">
                Plat.
                {service.realTimeUpdatesInfo.realTimeServiceInfo.realTimePlatform ||
                  service.scheduledInfo.scheduledPlatform}
              </div>
            )}
          {
            <div
              className={classNames('service-status', {
                ontime: isOnTime,
                delayed: !isOnTime
              })}
            >
              {isOnTime ? 'On Time' : isDelayed ? `Exp. ${formattedRealTime}` : isCancelled ? 'Cancelled' : 'Unknown'}
            </div>
          }
        </div>
      </StyledMiniService>
    );
  }
}

export default MiniService;
