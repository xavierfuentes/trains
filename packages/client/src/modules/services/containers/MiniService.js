import { connect } from 'react-redux';

import MiniService from '../components/MiniService';
import { activateService } from '../actions/services';

const mapStateToProps = state => ({
  active: state.services.active
});
const mapDispatchToProps = {
  activateService
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MiniService);
