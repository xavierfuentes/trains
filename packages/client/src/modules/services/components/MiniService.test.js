import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import MiniService from './MiniService';

describe('Test MiniService component', () => {
  const service = {
    serviceIdentifier: 'W13317',
    serviceOperator: 'SW',
    transportMode: 'TRAIN',
    scheduledInfo: { scheduledTime: '2018-05-29T13:06:00+01:00', scheduledPlatform: '2' },
    callingType: 'PickUp',
    destinationList: [{ crs: 'HMC' }],
    realTimeUpdatesInfo: {
      realTimeServiceInfo: {
        realTime: '2018-05-29T13:06:00+01:00',
        realTimePlatform: '2',
        realTimeFlag: 'Estimate'
      }
    },
    callingPatternUrl: 'https://realtime.thetrainline.com/callingPattern/W13317/2018-05-29'
  };

  it('renders correctly', () => {
    const tree = renderer.create(<MiniService service={service} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Tests click event', () => {
    const mockCallBack = jest.fn();

    const tree = shallow(<MiniService service={service} activateService={mockCallBack} />);
    tree.find('.service').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
