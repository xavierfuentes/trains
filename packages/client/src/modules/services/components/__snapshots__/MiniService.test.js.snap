// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MiniService component renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 0.071em;
  background: #ffffff;
  padding: 0.714em;
  border-left: 0.357em solid transparent;
}

.c0 .scheduled-time,
.c0 .service-info,
.c0 .service-realtime {
  padding: 0.357em;
}

.c0 .scheduled-time {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-weight: 700;
  font-size: 1.143em;
}

.c0 .service-info {
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
}

.c0 .service-realtime {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  text-align: right;
}

.c0 .service-operator {
  color: #9b9f9e;
  font-size: 0.857em;
}

.c0.active {
  background: #d8f3ec;
  border-left-color: #244774;
}

<article
  className="service c0"
  onClick={[Function]}
>
  <div
    className="scheduled-time"
  >
    13:06
  </div>
  <div
    className="service-info"
  >
    <div
      className="service-identifier"
    >
      W13317
    </div>
    <div
      className="service-operator"
    >
      SW
    </div>
  </div>
  <div
    className="service-realtime"
  >
    <div
      className="service-platform"
    >
      Plat.
      2
    </div>
    <div
      className="service-status ontime"
    >
      On Time
    </div>
  </div>
</article>
`;
